Allow values to be wrapped prior to serialization in etcd

smarterclayton opened the pull request:
This adds a new value transformer to the etcd2 store that can transform
the value from etcd on read and write. This will allow the store to
implement encryption at rest or otherwise transform the value prior to
persistence.

* [x] etcd3 store
* [x] example of transformation
* [x] partial error handling

This is in support of #12742

smarterclayton committed:
Allow values to be wrapped prior to serialization in etcd2

This adds a new value transformer to the etcd2 store that can transform
the value from etcd on read and write. This will allow the store to
implement encryption at rest or otherwise transform the value prior to
persistence.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40664)
<!-- Reviewable:end -->

smarterclayton commented:
@kubernetes/sig-api-machinery-pr-reviews

sttts commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -40,15 +40,29 @@ import (
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store.
+type ValueTransformer interface {
+	ReadString(string) (string, error)
```
Would prefer encode/decode. For Read/Write the signature is unexpected.

smarterclayton replied:
This isn't an encoding decoding step - it's a writer / reader wrapper.

smarterclayton replied:
I may make this ToStorage and FromStorage to be clearer.

sttts commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -498,6 +528,12 @@ func (h *etcdHelper) GuaranteedUpdate(
 		}
+		newBody := string(data)
+
+		body, err := h.transformer.WriteString(newBody)
```
A consistent naming of the transformed and raw value would it easier to read. E.g. body = transformed, value = raw, or conversely. Now it's mixed.

smarterclayton replied:
Yeah, I minimized the changes to the code.  I can make the names more explicit.

sttts commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -40,15 +40,29 @@ import (
 // Creates a new storage interface from the client
 // TODO: deprecate in favor of storage.Config abstraction over time
-func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quorum bool, cacheSize int) storage.Interface {
+func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quorum bool, cacheSize int, transformer ValueTransformer) storage.Interface {
```
Wouldn't it be simpler to decorate `etcdKeysAPI` with the transformation than adding ReadString/WriteString everywhere?

smarterclayton replied:
It's not going to work for etcd3 - no wrapping is possible today.

smarterclayton replied:
Not sure what you mean by key filtering?  This was not an ACL thing, this was intended to enable encrypt at rest for resources as an optional thing.  Definitely comes with a cost, needed to get to a measurable point.

From a use case point of view the simplest and least impactful is simple encrypt at rest, while others (encrypt before kube, external secret integration) are all reasonable and can be done out of tree, but are more opinionated.  Someone willing to pay the operational cost for encrypt at rest can do so in tree.

timothysc replied:
K, knowing the motivation now makes more sense.. The original linked issue points out secrets explicitly.  

One thing I'm struggling with in this WIP is why we wouldn't put this behavior into the codec and serialization f(n)s of encode and decode?  As it seems like a natural fit...

timothysc commented:
Not really a huge fan of this approach.  It seems like this capability should be able to be built into the client itself.  

/cc @hongchaodeng @xiang90

smarterclayton commented:
I explored that but it's not feasible in the etcd3 client as is.  Also, I don't think wrapping an entire client api (including all the issues w.r.t. body compares) is appropriate when the thing being transformed is just a different step.  Also, there's no guarantee a new storage author will be able to wrap their client api.

This originally started in encode/decode, but that doesn't have the right level of visibility to the store for GuaranteedUpdate in order to ensure that idempotence of updates is preserved (we don't want to reuse nonces).

Within the limitations we have, transformation of the body in the store is as close as we get to the actual requirement.

smarterclayton commented:
Basically, there are three things that look similar but are not:

1. Encoding / decoding wrappers work when the transformation of encoded data is stable - encryption of data at rest is not necessarily stable because we do not want / need to reuse nonces.
2. Wrapping client libraries means you have to wrap the entire surface area of the API and the caller may miss spots - also not all client libraries are wrappable
3. Altering the data at rest *outside* of the semantics of the API requires integration into the storage label.

timothysc commented:
If the goal is to address #12742, I don't see a reason why the capability couldn't be in the registry get/puts on the secrets resource.

liggitt commented:
> I don't see a reason why the capability couldn't be in the registry get/puts on the secrets resource.

has to be in the storage layer because of all the plumbing below the generic registry (watch cache, guaranteedupdate, etc)

deads2k commented:
It seems pretty reasonable to me.  Complicated, potentially one way, but reasonable.  Have you considered what kind of error the API produces if it can't untransform?

deads2k commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -113,7 +128,13 @@ func (h *etcdHelper) Create(ctx context.Context, key string, obj, out runtime.Ob
+
+	value, err := h.transformer.WriteString(string(data))
+	if err != nil {
+		return err
```
I'd like a typed error.

deads2k commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -283,7 +304,11 @@ func (h *etcdHelper) extractObj(response *etcd.Response, inErr error, objPtr run
+
+	body, err = h.transformer.ReadString(node.Value)
+	if err != nil {
+		return body, nil, err
```
I think we need a typed error here so that the API can report the issue back to the client which can decide how to proceed.

smarterclayton replied:
Yeah, was going to be another commit in this PR once I got the basic "ew, gross" reactions out of the way.

liggitt commented:
> Have you considered what kind of error the API produces if it can't untransform?

do we still fail a list if any one item cannot be decoded? this makes it far more likely we'll hit a scenario like that

smarterclayton commented:
> do we still fail a list if any one item cannot be decoded? 

Yes, we do.  I was debating how we would handle that.  I think we're going to have to do partial decoding and return the right errors on update / create / delete.

smarterclayton committed:
Allow values to be wrapped prior to serialization in etcd3

smarterclayton commented:
@k8s-bot kops aws e2e test this

smarterclayton commented:
Added example of errors and a test to etcd2 path.  Get, Delete, and Create error.  Delete succeeds (decoding body is not required to delete).  List elides results that have errors and logs them to HandleError, and Watch already does that.  Used the storage.InternalError which seems most closely related to "admin fat fingered the security credentials".

sttts commented on pkg/storage/etcd3/store.go:
```diff
@@ -39,15 +39,33 @@ import (
+	// TransformStringFromStorage may transform the provided string from its underlying storage representation or return an error.
+	TransformFromStorage([]byte) ([]byte, error)
+	// TransformStringToStorage may transform the provided string into the appropriate form in storage or return an error.
+	TransformToStorage([]byte) ([]byte, error)
```
This is clearer! 👍

sttts commented on pkg/storage/etcd/etcd_watcher.go:
```diff
@@ -71,8 +71,9 @@ type etcdWatcher struct {
-	versioner storage.Versioner
-	transform TransformFunc
+	versioner        storage.Versioner
+	transform        TransformFunc
```
a short comment wouldn't harm, now that we have two transforming concepts.

sttts commented on pkg/storage/etcd/etcd_helper.go:
```diff
@@ -40,15 +41,33 @@ import (
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
+	// TransformStringFromStorage may transform the provided string from its underlying storage representation or return an error.
+	TransformStringFromStorage(string) (string, error)
```
No need for the `String` in the func names.

smarterclayton replied:
I want to ensure that one implementation can satisfy both interfaces (a decrypter) so the names need to be different.

sttts commented on pkg/storage/etcd3/store.go:
```diff
@@ -39,15 +39,33 @@ import (
 	"k8s.io/kubernetes/pkg/storage/etcd"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
```
a `[]byte` value

sttts commented on pkg/storage/etcd3/store.go:
```diff
@@ -39,15 +39,33 @@ import (
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
+	// TransformStringFromStorage may transform the provided string from its underlying storage representation or return an error.
```
same here

sttts commented:
I haven't seen an actual encryption transformer implementation yet, but here is a remark:

There is one assumption we cannot make about encryption: it will not return the same encrypted data for the same input. The reason is it will use some random padding behind the actual input to avoid certain attacks. In other words: `encrypt(data)` will change over time. 

This might rule out certain alternative layers to put the encryption in. 

@liggitt can you comment on this, you seem to refer to something similar in https://github.com/kubernetes/kubernetes/pull/40664#issuecomment-276147542

smarterclayton commented:
Yes, that is why this layer exists.  Because GuaranteedUpdate must be able to compare API bytes -> API bytes for change, and you cannot compare encrypted data like that.

smarterclayton commented:
This is feature complete (for the intended purpose) - tests and error handlers.

smarterclayton commented:
@k8s-bot verify test this

smarterclayton commented:
@k8s-bot you're drunk, go home.

smarterclayton committed:
Allow ValueTransformer to indicate resource is stale

Allows a transformer (such as an encrypter) to force an update if a new
key is in use, thus allowing simple writes to the REST layer to
trivially migrate keys.

smarterclayton commented:
Updated with a new boolean (and other fixes on the PR) that allow the Transform layer to signal to GuaranteedUpdate that the data is stale and thus an update *should* be done even if the data hasn't changed.  That would allow keys to be transformed.  This is another reason why Encode wrappers are insufficient and a client wrapper would also be insufficient - GuaranteedUpdate itself is required.

This is complete for the needs of the encrypting store I think.

timothysc commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -33,21 +33,42 @@ import (
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
```
Is the only reason this isn't abstracted is because of byte[] vs string?   Can't we de-dupe?

smarterclayton replied:
I was hoping we'd drop etcd2 first.  For now, I'd like to avoid unnecessary copies on every store (etcd3 is forcing it), but I would expect that someone would implement both interfaces as efficiently as possible.

timothysc reviewed (commented):
Only minor comment.  

@hongchaodeng or @wojtek-t any comments?

smarterclayton commented:
Any other comments?
